### PR TITLE
expression: implement vectorized evaluation for 'builtinRandWithSeedSig'

### PR DESCRIPTION
**Pull request by tsthght**
### What problem does this PR solve? 
expression: implement vectorized evaluation for 'builtinRandWithSeedSig', for #12105

### What is changed and how it works?
```
BenchmarkVectorizedBuiltinMathFunc/builtinRandWithSeedSig-VecBuiltinFunc-8         	     100	  10981342 ns/op	 5554210 B/op	    2048 allocs/op
BenchmarkVectorizedBuiltinMathFunc/builtinRandWithSeedSig-NonVecBuiltinFunc-8      	     100	  11391656 ns/op	 5505037 B/op	    1024 allocs/op
```

### Check List
Returns random numbers, so cannot write test cases

### CODE REVIEW

**Comment by tsthght** [comment]
> Please do a benchmark for it and post the result in the description.
> For correct testing, let's discuss in this [PR](https://github.com/pingcap/tidb/pull/12656).

ok

File: expression/builtin_math_vec.go [hunk 1]
```diff
@@ -687,11 +689,33 @@ func (b *builtinRandSig) vecEvalReal(input *chunk.Chunk, result *chunk.Column) e
...
 
 func (b *builtinRandWithSeedSig) vecEvalReal(input *chunk.Chunk, result *chunk.Column) error {
-	return errors.Errorf("not implemented")
+	n := input.NumRows()
+	buf, err := b.bufAllocator.get(types.ETInt, n)
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(buf)
+	if err := b.args[0].VecEvalInt(b.ctx, input, buf); err != nil {
```
Comment (XuHuaiyu) [inline expression/builtin_math_vec.go]: s/ `err :=`/ `err =`

Comment (tsthght) [reply expression/builtin_math_vec.go]: thank you, I had modified.

File: expression/builtin_math_vec.go [hunk 2]
```diff
@@ -687,11 +689,33 @@ func (b *builtinRandSig) vecEvalReal(input *chunk.Chunk, result *chunk.Column) e
...
+		return err
+	}
+
+	result.ResizeFloat64(n, false)
+	var randGen *rand.Rand
+	i64s := buf.Int64s()
+	f64s := result.Float64s()
+	for i := 0; i < n; i++ {
+		if buf.IsNull(i) {
+			randGen = rand.New(rand.NewSource(time.Now().UnixNano()))
```
Comment (qw4990) [inline expression/builtin_math_vec.go]: How about creating one rander outside the loop for all nulls?

Comment (tsthght) [reply expression/builtin_math_vec.go]: Did you mean like that ?
```
	rander := rand.NewSource(time.Now().UnixNano())
	for i := 0; i < n; i++ {
		if buf.IsNull(i) {
			randGen = rand.New(rander)
		} else {
			randGen = rand.New(rand.NewSource(i64s[i]))
		}
		f64s[i] = randGen.Float64()
	}
```
I had modified the code.

Comment (tsthght) [reply expression/builtin_math_vec.go]: @qw4990  PTAL

File: expression/builtin_math_vec.go [hunk 3]
```diff
@@ -687,11 +689,34 @@ func (b *builtinRandSig) vecEvalReal(input *chunk.Chunk, result *chunk.Column) e
...
+	result.ResizeFloat64(n, false)
+	var randGen *rand.Rand
+	i64s := buf.Int64s()
+	f64s := result.Float64s()
+	rander := rand.NewSource(time.Now().UnixNano())
+	for i := 0; i < n; i++ {
+		if buf.IsNull(i) {
+			randGen = rand.New(rander)
+		} else {
+			randGen = rand.New(rand.NewSource(i64s[i]))
```
Comment (SunRunAway) [inline expression/builtin_math_vec.go]: We can keep the result of `rand.New` outside the fop loop and use https://godoc.org/math/rand#Rand.Seed here.

Comment (tsthght) [reply expression/builtin_math_vec.go]: OK， I will try it.

Comment (tsthght) [reply expression/builtin_math_vec.go]: I had Modified  the implementation of both non-vec and vec. PTAL @SunRunAway

**Comment by tsthght** [comment]
> Please consider write a separated test.

I had Modified the implementation of both non-vec and vec. I don't know whether it's right. PTAL @SunRunAway @qw4990

File: expression/builtin_math.go [hunk 4]
```diff
@@ -1028,10 +1028,8 @@ func (b *builtinRandWithSeedSig) evalReal(row chunk.Row) (float64, bool, error)
 	// b.args[0] is promised to be a non-constant(such as a column name) in
 	// builtinRandWithSeedSig, the seed is initialized with the value for each
 	// invocation of RAND().
-	var randGen *rand.Rand
-	if isNull {
-		randGen = rand.New(rand.NewSource(time.Now().UnixNano()))
-	} else {
+	randGen := rand.New(rand.NewSource(time.Now().UnixNano()))
+	if !isNull {
 		randGen = rand.New(rand.NewSource(seed))
```
Comment (qw4990) [inline expression/builtin_math.go]: Please keep the `else` branch here then we can avoid creating an unnecessary rander if there is a seed.

Comment (tsthght) [reply expression/builtin_math.go]: I had modified.

**Pull request merged**